File: cpp/StateDemo.cpp
#include "SingleTon.h"

#include <stdio.h>
#include <unistd.h>
#include <string>
#include <iostream>

using namespace std;

class TCPState;

class TCPConnetion
{
	public:
		TCPConnetion();
		~TCPConnetion() {}

		//used for client
		void ActiveOpen();
		//used for server
		void PassiveOpen();
		void Close();

		void Send(string& msg);
		void Recv(const string& msg);

	private:
		friend class TCPState;
		void ChangeState(TCPState*);

	private:
		TCPState* state_;
};

class TCPState
{
	public:
		virtual void ActiveOpen(TCPConnetion* ) {}
		virtual void PassiveOpen(TCPConnetion* ) {}
		virtual void Close(TCPConnetion*)	{}
		virtual void Send(TCPConnetion*, string& msg) {}
		virtual void Recv(TCPConnetion*, const string& msg) {}

		virtual ~TCPState() = default;
		virtual const string& getStateName() const = 0;

	protected:
		void ChangeState(TCPConnetion*, TCPState*);
};

class TCPListen: public TCPState, public SingleTon<TCPListen>
{
	public:
		void Send(TCPConnetion* c, string& msg) override;
		void Recv(TCPConnetion* c, const string& msg) override;
		const string& getStateName() const override {
			return name_;
		}
	private:
		const string name_ = "LISTEN";
};

class TCPEstablished : public TCPState, public SingleTon<TCPEstablished>
{
	public:
		const string& getStateName() const override{
			return name_;
		}
	private:
		const string name_ = "ESTABLISHED";
};

class TCPClosed : public TCPState, public SingleTon<TCPClosed>
{
	public:
		void ActiveOpen(TCPConnetion* c) override {
			ChangeState(c, &TCPEstablished::getInstance());
		}

		void PassiveOpen(TCPConnetion* c) override {
			ChangeState(c, &TCPListen::getInstance());
		}

		const string& getStateName() const override{
			return name_;
		}
	private:
		const string name_ = "CLOSED";

};

class TCPSynReceived: public TCPState, public SingleTon<TCPSynReceived> 
{
	public:
		void Recv(TCPConnetion* c, const string& msg) override {
			if (msg=="ACK")	 {
				ChangeState(c, &TCPEstablished::getInstance());
			}
		}

		const string& getStateName() const override {
			return name_;
		}
	private:
		const string name_ = "SYN-RECEIVED";
};

void TCPListen::Send(TCPConnetion* t, string& msg)
{
	cout<<"Send "<<msg<<endl;
	ChangeState(t, &TCPSynReceived::getInstance());
}

void TCPListen::Recv(TCPConnetion* c, const string& msg) {
	if (msg == "ACK") {
		string msg = "SYN+ACK";
		Send(c, msg);
	}
}

TCPConnetion::TCPConnetion()
{
	state_ = &TCPClosed::getInstance();
	cout<<"TCPConnetion init state "<<state_->getStateName()<<endl;
}

void TCPConnetion::ChangeState(TCPState* s)
{
	state_ = s;
}

void TCPConnetion::ActiveOpen()
{
	state_->ActiveOpen(this);
}

void TCPConnetion::PassiveOpen()
{
	cout<<"TCPConnetion setup tcb"<<endl;
	state_->PassiveOpen(this);
}

void TCPConnetion::Close()
{
	state_->Close(this);
}

void TCPConnetion::Send(string& msg)
{
	state_->Send(this, msg);
}

void TCPConnetion::Recv(const string& msg)
{
	cout<<"TCPConnetion Recvd [ "<<msg<<" ]"<<endl;
	state_->Recv(this, msg);
}

void TCPState::ChangeState(TCPConnetion* t, TCPState* s)
{
	cout<<"[ "<<getStateName()<<" ] goto "<<"[ "<<s->getStateName()<<"] "<<endl<<endl;
	t->ChangeState(s);
}

int main(int argc, char *argv[])
{
	TCPConnetion c;
	c.PassiveOpen();
	c.Recv("ACK");	
	c.Recv("ACK");


	return 0;
}

File: java/behavioral/StateDemo.java
package behavioral;

//tcp连接，持有tcp状态机
class TCPConnetion {
    TCPConnetion() {
        this.state = new TCPClosed();
        System.out.println("TCPConnetion init state " + state.getStateName());
    }

    void PassiveOpen() {
        System.out.println("TCPConnetion setup tcb");
        this.state.PassiveOpen(this);
    }

    void Close() {
        state.Close(this);
    }

    void Send(String msg) {
        state.Send(this, msg);
    }

    void Recv(String msg) {
        System.out.println("TCPConnetion Recvd [ " + msg + " ]");
        state.Recv(this, msg);
    }

    void ChangeState(TCPState s) {
        this.state = s;
    }

    private TCPState state;
};

//tcp状态机基类，负责处理特定状态下的各种事件
//每一个接口，表示对一个事件的处理
abstract class TCPState {
    protected TCPState(String name) {
        this.name = name;
    }

    //主动打开连接， 用于客户端
    void ActiveOpen(TCPConnetion c) {
    }

    //被动打开连接，用于服务端
    void PassiveOpen(TCPConnetion c) {
    }

    //关闭连接
    void Close(TCPConnetion c) {

    }

    //发出报文
    void Send(TCPConnetion c, String msg) {

    }

    //收到报文
    void Recv(TCPConnetion c, String msg) {
    }

    //获取当前状态名称
    String getStateName() {
        return name;
    }

    void ChangeState(TCPConnetion c, TCPState s) {
        System.out.println("[" + getStateName() + "] goto [" + s.getStateName() + "] \n");
        c.ChangeState(s);
    }

    private String name;
};

class TCPListen extends TCPState {
    TCPListen() {
        super("LISTEN");
    }

    @Override
    void Send(TCPConnetion c, String msg) {
        System.out.println("Send " + msg);
        ChangeState(c, new TCPSynReceived());
    }

    @Override
    void Recv(TCPConnetion c, String msg) {
        if (msg == "ACK") {
            Send(c, "SYN+ACK");
        }
    }
};

class TCPEstablished extends TCPState {
    TCPEstablished() {
        super("ESTABLISHED");
    }
};

class TCPClosed extends TCPState {
    TCPClosed() {
        super("CLOSED");
    }

    @Override
    void PassiveOpen(TCPConnetion c) {
        ChangeState(c, new TCPListen());
    }
};

class TCPSynReceived extends TCPState {
    TCPSynReceived() {
        super("SYN-RECEIVED");
    }

    @Override
    void Recv(TCPConnetion c, String msg) {
        if (msg == "ACK") {
            ChangeState(c, new TCPEstablished());
        }
    }
};

public class StateDemo {
    public static void main(String[] args) {
        TCPConnetion c = new TCPConnetion();
        c.PassiveOpen();    //tcp连接进入listen状态
        c.Recv("ACK"); //收到第一步握手的ack
        c.Recv("ACK"); //收到第三步握手的ack，连接建立成功
    }
}
